monkestation/code/modules/antagonists/contractor/items/misc.dm

```diff
@@ -7,6 +7,111 @@
+	icon_state = "pinpointer_syndicate"
+	worn_icon_state = "pinpointer_black"
+	special_examine = TRUE
+	var/area/tracked_area
+	var/pinpointer_owner
+	/// We save our position every time we scan_for_target()
+	/// Its used to check if we moved so we may ignore calculations when being still, along with calculations between us and the target turfs.
+	var/turf/pinpointer_turf
+
+/obj/item/pinpointer/area_pinpointer/Destroy()
+	tracked_area = null
+	pinpointer_turf = null
+	return ..()
+
+// we need to get our own examine text, since it would be "tracking the floor" otherwise
+/obj/item/pinpointer/area_pinpointer/examine(mob/user)
+	. = ..()
+	if(target)
+		. += "It is currently tracking [tracked_area]."
+
+/obj/item/pinpointer/area_pinpointer/get_direction_icon(here, there)
+	var/list/area_turfs = list()
+	area_turfs = get_area_turfs(tracked_area)
+
+	var/turf/pinpointer_turf = get_turf(src)
+	for(var/turf/possible_turf as anything in area_turfs) // im a bit scared of the effects this may have on performance, but it seems fine
+		if(get_dist_euclidian(pinpointer_turf, possible_turf) <= minimum_range)
+			return "pinon[alert ? "alert" : ""]direct[icon_suffix]"
+
+	return ..()
+
+/obj/item/pinpointer/area_pinpointer/scan_for_target()
+	var/current_turf = get_turf(src)
+
+	if(pinpointer_turf == current_turf) // if our position has not changed, we dont need to update our target.
+		return
+
+	pinpointer_turf = current_turf
+
+	var/list/area_turfs = list()
+	area_turfs = get_area_turfs(tracked_area)
+
+	/// The turf that has the lowest possible range towards us and the area
+	var/turf/closest_turf
+	/// Whats the range between us and the closest turf?
+	var/closest_turf_range = 255
+	for(var/turf/floor as anything in area_turfs) // Lets go over everything and check their distances for the closest tile, what can go wrong?
+		if(floor.density) // lets ignore all walls before getting every turfs distance, to save some performance
+			continue
+		if(get_dist_euclidian(pinpointer_turf, floor) < closest_turf_range)
+			closest_turf_range = get_dist_euclidian(pinpointer_turf, floor)
+			closest_turf = floor
+
+	target = closest_turf
+
+/obj/item/pinpointer/area_pinpointer/attack_self(mob/living/user)
+	if(active)
+		toggle_on()
+		user.visible_message(span_notice("[user] deactivates [user.p_their()] pinpointer."), span_notice("You deactivate your pinpointer."))
+		return
+
+	if(!pinpointer_owner)
+		pinpointer_owner = user
+
+	if(pinpointer_owner != user)
+		to_chat(user, span_notice("The pinpointer doesn't respond. It seems to only recognise its owner."))
+		return
+
+	// This list should ONLY include areas that are on our z-level and are actually recognizable, else we confuse the contractor
+	var/list/possible_areas = list()
+	for(var/area/area in GLOB.areas)
+		var/our_z = user?.z
+		var/area_z = area?.z
```

Comment:
the ?. here is not needed as if area is null something has gone TERRIBLY wrong somewhere else and this is gonna give a misleading error message, also with the !user check at the top the ?. for user is also redundant